google/resource_container_node_pool.go

```diff
@@ -85,6 +85,29 @@ var schemaNodePool = map[string]*schema.Schema{
 		Deprecated: "Use node_count instead",
 	},
 
+	"management": {
+		Type:     schema.TypeList,
+		Optional: true,
+		Computed: true,
```

Comment:
Is this actually computed? (note: it might be! it's just sometimes hard to tell from looking at the api, so I usually ask)

Reply:
I'm obviously somewhat fuzzy on the exact meaning of `computed` and when exactly to use it / not use it (especially in the context of new attributes with `ForceNew: true` being added to resources that are already out in the wild, though this is not that case), if you don't mind giving me a one-time clarification.  In this context, the field is always passed back from the API and both values are false if unspecified.  Based on my current understanding, I'd think that means the field is `optional` + `computed`, but that could be totally incorrect.  I can do a bit of testing on my own to get a better grasp.

Reply:
The way I like to think about `Computed` is that it doesn't cause a diff if the value is set in state but not in the config. For a simple case, take `name`. We allow the user to set this field to empty, which means that terraform will generate a name for you, which we then store in state. Since we don't want the next run of `terraform plan` to show a diff (since you aren't _actually_ trying to set the name to empty), we have `name` as `Computed` so that when it's empty, nothing happens.

So in this case, if management is always returned from the api, then it will always get set in state via the flatten fn, so `Computed` is indeed the correct choice. A good example of the opposite is autoscaling- if it isn't enabled, then it doesn't get returned from the API at all (so `np.Autoscaling` gives nil), meaning we don't want it to get set in state.

The reason I like to double-check that a field needs to be`Computed` is that this empty-means-no-diff-behavior means that removing the field from your config doesn't set it to the default value, as might be expected. So for example, if we completely remove the `autoscaling` block, terraform will notice there's a diff between what's in state (autoscaling enabled) and what's in the config (no autoscaling) and will call update appropriately. For management, if (say) `auto_repair` is set to true in state, and then I remove the entire `management` block in the hopes of "removing" management, terraform won't see that as a diff and so nothing will be updated.

The way I usually handle this when developing is to just set nothing to Computed and see what fails in testing. In this particular case, since `Management` is returned from the API, I think you should leave it as `Computed` but set the two properties to be `Default: false`. That way, if they are set to true in the schema and then removed, terraform will see that it needs to set it back to false. It'll just work for those individual fields, not the full management block, but I think it's still better than nothing.

I hope that makes sense! It's taken me a long time to get to my current level of understanding so feel free to ask any more questions you might have :)

Reply:
Rad!  Thanks so much for the great, thorough explanation.  That makes complete sense.  I'll make that change and re-test now.